### PR TITLE
Fix to handle NULL value with Informix ODBC Driver

### PR DESCRIPTION
**Pull request by lmtyler**

### CODE REVIEW

**Comment by lmtyler** [comment]
Here is an example of the data.

`&{0xc0000b00e0 true true 31 4294967295 [80 79 32 66 79 88 32 52 50 56 32 32 32 32 32 32 32 32 32 32 32 32 32 32 32 32 32 32 32 32 0]}`

File: column.go [hunk 1]
```diff
@@ -17,7 +17,7 @@ import (
 type BufferLen api.SQLLEN
 
 func (l *BufferLen) IsNull() bool {
-	return *l == api.SQL_NULL_DATA
+	return int16(*l) == api.SQL_NULL_DATA
```
Comment (alexbrainman) [inline column.go]: Sorry, but I don't see why this change is needed. What is the problem that this change fixes?

Alex

Comment (lmtyler) [reply column.go]: -1 is a two's complement thus typing the value to a smaller memory space does not change the value.

Unfortunately, when using the 64bit IBM Informix ODBC driver on Linux they use only 8 bytes of memory for the NULL value indicator , thus when put into a standard Go int (long long) it does not fill the memory and you get value 4,294,967,295 and not a -1.

Since the other drivers you have tested with fill the whole memory space there should be no adverse affects.

BTW, this should also allow for the usage with 32bit builds as well, but honestly it might need to be int8 for 32bit.

Comment (alexbrainman) [reply column.go]: > -1 is a two's complement thus typing the value to a smaller memory space does not change the value.

Sorry, but I do not understand your explanation.

> Unfortunately, when using the 64bit IBM Informix ODBC driver on Linux they use only 8 bytes of memory for the NULL value indicator

OK.

> thus when put into a standard Go int (long long)

Go int on amd64 is int64.

> it does not fill the memory and you get value 4,294,967,295 and not a -1.

The value is also determined by the int type. If memory is defined as int32, you will get -1. If memory is defined as uint32, you will get 4294967295. But I don't see how this is relevant to your change.

Alex

Comment (alexbrainman) [reply column.go]: > For a NULL indicator we get back an int32(-1) which is 0x00000000FFFFFFFF ...

If you are talking about SQL_NULL_DATA, I don't see SQL_NULL_DATA converted to int32 here.
```
$ grep SQL_NULL_DATA *
sql.h:#define SQL_NULL_DATA             (-1)
$
```
I can see Cgo represents SQL_NULL_DATA as int.
```
$ git diff
diff --git a/mssql_test.go b/mssql_test.go
index 5eb10fa..ca75fc4 100644
--- a/mssql_test.go
+++ b/mssql_test.go
@@ -1732,3 +1732,9 @@ func TestMSSQLMarkBeginBadConn(t *testing.T) {
                testFn(next.label, next.fn)
        }
 }
+
+func TestALEX(t *testing.T) {
+       t.Logf("%T %v", api.SQL_NULL_DATA, api.SQL_NULL_DATA)
+       t.Logf("%v", int32(api.SQL_NULL_DATA))
+       t.Logf("%v", int64(api.SQL_NULL_DATA))
+}
$ go test -run=ALEX -v
=== RUN   TestALEX
--- PASS: TestALEX (0.00s)
    mssql_test.go:1737: int -1
    mssql_test.go:1738: -1
    mssql_test.go:1739: -1
PASS
ok      github.com/alexbrainman/odbc    0.004s
$
```

Alex

Comment (lmtyler) [reply column.go]: Sorry Alex, I guess I am not being clear.

It is not an issue with sql.h nor the original code directly, but with the data that is populating BaseColumn being evaluated by IsNull.

Back to my original comment:
>&{0xc0000b00e0 true true 31 4294967295 [80 79 32 66 79 88 32 52 50 56 32 32 32 32 32 32 32 32 32 32 32 32 32 32 32 32 32 32 32 32 0]}

This is the data returned for sqlNullString from the 64bit Informix ODBC Driver thru the unixODBC Manager.

Notice the **4294967295**, that is the value coming back in Len of BindableColumn that is being passed into IsNull
https://github.com/alexbrainman/odbc/blob/master/column.go#L252

Comment (alexbrainman) [reply column.go]: > Back to my original comment:
> 
> > &{0xc0000b00e0 true true 31 4294967295 [80 79 32 66 79 88 32 52 50 56 32 32 32 32 32 32 32 32 32 32 32 32 32 32 32 32 32 32 32 32 0]}
> 
> This is the data returned for NULL string defined as sql.NullString from the 64bit Informix ODBC Driver thru the unixODBC Manager.
> 
> Notice the **4294967295**, that is the value coming back in Len of BindableColumn

Sure. But Len of BindableColumn type is BufferLen, and BufferLen type is api.SQLLEN, and api.SQLLEN type is C.SQLLEN. And api.SQLLEN on my computer is 8 bytes long.
```
$ git diff
diff --git a/mssql_test.go b/mssql_test.go
index 5eb10fa..83d2747 100644
--- a/mssql_test.go
+++ b/mssql_test.go
@@ -20,6 +20,7 @@ import (
        "sync/atomic"
        "testing"
        "time"
+       "unsafe"
 
        "github.com/alexbrainman/odbc/api"
 )
@@ -1732,3 +1733,11 @@ func TestMSSQLMarkBeginBadConn(t *testing.T) {
                testFn(next.label, next.fn)
        }
 }
+
+func TestALEX(t *testing.T) {
+       var l api.SQLLEN
+       t.Logf("SQLLEN size is %v", unsafe.Sizeof(l))
+       l--
+       t.Logf("SQLLEN -1 is %v", l)
+       t.Logf("l == api.SQL_NULL_DATA is %v", l == api.SQL_NULL_DATA)
+}
$ go test -run=ALEX -v
=== RUN   TestALEX
--- PASS: TestALEX (0.00s)
    mssql_test.go:1739: SQLLEN size is 8
    mssql_test.go:1741: SQLLEN -1 is -1
    mssql_test.go:1742: l == api.SQL_NULL_DATA is true
PASS
ok      github.com/alexbrainman/odbc    0.004s
$ 
```
And -1 of api.SQLLEN is not 4294967295, but much larger. 4294967295 is uint32 equivalent of int32(-1) - so your C.SQLLEN must be 4 bytes long. My SQLLEN is defined in sqltypes.h - I can see SQLLEN can be either 4 or 8 bytes long - depending on your configuration. You need to understand why your SQLLEN is 4 bytes long on amd64.

Alex

Comment (lmtyler) [reply column.go]: Alex,
As I stated
> NOTE: I am also driving this back thru our IBM reps as a bug in their driver, but have no idea when or even if they will except it as a bug to get corrected

#### Still working on this
Unfortunately, the repos have already said that is it unlikely that this will be acknowledged as a bug, and even if it would be acknowledged as a bug it would only be fixed it later versions for the Informix ODBC drivers.  

But this change:

1. Will allow for the code to support a 32bit driver running on amd64
2. Support anything equal to or smaller than 2147483647 in length from a single column

Comment (alexbrainman) [reply column.go]: > But this change:
> 
> 1. Will allow for the code to support a 32bit driver running on amd64
> 2. Support anything equal to or smaller than 2147483647 in length from a single column

But your change assumes that only lowest 2 bytes of api.SQLLEN integer are important - you ignore everything else. But your assumption is not documented anywhere. Do you have a link to some documentation about that? What will happen to people on the systems where your assumption does not hold true? They programs will fail.

I cannot accept your change. If it works for you, then use it. Why break other people's programs?

Alex

**Comment by lmtyler** [comment]
Understood

**Comment by alexbrainman** [comment]
> Understood

Your last commit changes all import path of `github.com/alexbrainman/odbc/api` for `github.com/lmtyler/odbc/api`, and that is fine, if you want to submit it yourself in your own Github repo. I won't accept it here - otherwise this package will not build.

Alex

**Comment by lmtyler** [comment]
Yes I am no longer going to PR this. I am going to use mine for now and drive it back to IBM Informix as an issue in their driver.

**Comment by alexbrainman** [comment]
> Yes I am no longer going to PR this.

Sounds good to me. Please close this PR, if it is not needed.

Thank you.

Alex
